sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoder.java

```diff
@@ -136,6 +137,20 @@ private void verifyDeterministic(Schema schema)
     Coder.verifyDeterministic(this, "All fields must have deterministic encoding", coders);
   }
 
+  @Override
+  public Object structuralValue(final Row row) {
```

Comment:
I think making `Row.equals` a custom deep equality (so byte arrays with the same contents are equal, etc) is the best solution here. Then it is always consistent with equals and no structuralValue is needed. Is there a pitfall?

Reply:
That doesn't really make sense. The comparisons you have to do for correctness are the same, and what I'm suggestion would eliminate the copying/allocation.

Reply:
Row still has incorrect hashcode and equals so DirectRunner would actually just shuffle wrong if there were any BYTES field. Which runner was it?